### ISSUE TITLE
Pairs trade opening priority

### ISSUE THREAD

**Issue by denmoroz** [freqtrade/freqtrade#6030]
## Describe your environment
(if applicable)

  * Operating system: ____
  * Python Version: _____ (`python -V`)
  * CCXT version: _____ (`pip freeze | grep ccxt`)
  * Freqtrade Version: ____ (`freqtrade -V` or `docker-compose run --rm freqtrade -V` for Freqtrade running in docker)

## Describe the enhancement

Sometimes it could be useful to order pairs **after** dataframe analysis for each pair is done (probably based on some indicator or whatever).

Let's imagine that target indicator is "probability of profit" for a pair during the following K candles, for instance predicted by some Machine Learning technique. Someone may want to open trades in max-to-min predicted probability order to maximize expected wins / by expected return based on predicted probability and position size to maximize expected capital growth / whatever else.

Currently there is no **clear way** to provide "priority" information out of the strategy so the bot will know which pair to enter first. One approach I can imagine right now is to override `analyze(self, pairs: List[str])` somehow like:
```
from queue import PriorityQueue

def analyze(self, pairs: List[str]) -> None:
    priority_queue = PriorityQueue()

    while pairs:
        pair = pairs.pop(0)
        pair_priority = 0.0

        self.analyze_pair(pair)
        analyzed_df, _ = self.dp.get_analyzed_dataframe(pair=pair, timeframe=self.timeframe)

        if not analyzed_df.empty:
            latest_candle = analyzed_df.iloc[-1].squeeze()
            pair_priority = latest_candle[PRIORITY_SIGNAL_COLUMN]

        priority_queue.put((pair_priority, pair))

    while not priority_queue.empty():
        best_priority, best_pair = priority_queue.get()
        pairs.append(best_pair)
```

or just simply `.sort()` pairs list inplace based on some column calculated in `populate_indicators`. But is definitely **not** a straightforward and potentially dangerous approach as it is based on knowledge of bot cycle [internals](https://github.com/freqtrade/freqtrade/blob/2021.11/freqtrade/freqtradebot.py#L167-L183).

What I imagine could be useful is change the `IStrategy` interface so there will be a point to explicitly provide priorities for the bot cycle and order pairs before [enter_positions](https://github.com/freqtrade/freqtrade/blob/2021.11/freqtrade/freqtradebot.py#L181-L183). This method could be optional to override with default return value of 0.0 so by default there will not be any specific ordering and everything will work exactly how it works right now.

**Comment by xmatthias** [comment]
This is definitely not a supported way, and i'd discourage you (or any future reader) from pursuing it, as it can break your strategy in subtle ways during updates (or completely ...).

It will also not work as you think it does - as you're modifying the order the pairs are analyzed (which is irrelevant anyway, and cannot be guaranteed on "candle time switches")- but not the order of the pairlist itself.

---

Unfortunately however, messing with the pairlist (in a proper way) is not an easy task / approach.
What you're proposing does only fit your particular need, but from a framework perspective, would not make much sense as it'd be quite limiting (and we'd need to have quite strict / complex logic in place to ensure that the pairlist is not modified in a breaking manner).

if we would want to enable this, then it would need to be called at the level of pairlists (where pairlists are populated) - also allowing users to provide new pairs from within the strategy.
If that were the case however, then you could not use any technical indicators (freqtrade can't provide the dataframe, and users can't rely on the dataframe being available through dataprovider).

As such, i think this task is better left to dedicated pairlists plugin.
It might be possible to simplify adding pairlist plugins (#4477) - but for the moment, i certainly don't see us having a callback in the strategy to sort the pairlist.

**Comment by denmoroz** [comment]
@xmatthias Thanks for the answer.

> I certainly don't see us having a callback in the strategy to sort the pairlist.

It worth to say that I didn't mean exactly this. I meant that strategy might provide some optional "importance" signal along with "buy" signal. It may be even indicator itself - let's say RSI value (the lower RSI the more oversold the asset assumed to be and presumably more desired to buy for RSI-based strategy). It might be useful when many pairs generate "buy" signal simultaneously but the amount of free stake currency is not enough to enter them all - then it is probably worth to be greedy and enter most "important" / promising first.

And then, **the bot** will be able to order analyzed pairs based on their importance in bot's main loop (not the strategy). If strategy does not provide any "importance" signal then there is no any preferred ordering for trades opening.

**Comment by stash86** [comment]
A little bit late, but this would do what you want. In this example, pairs are sorted based of their RSI values. And this snippet works on backtest as well.

https://botacademy.ddns.net/2024/06/26/how-to-trade-only-x-amount-of-pairs-with-lowest-highest-y-indicator-value/

**Comment by malone6** [comment]
> A little bit late, but this would do what you want. In this example, pairs are sorted based of their RSI values. And this snippet works on backtest as well.
> 
> https://botacademy.ddns.net/2024/06/26/how-to-trade-only-x-amount-of-pairs-with-lowest-highest-y-indicator-value/

this snippet is not working? In backtesting, `confirm_trade_entry` calls `self.dp.get_analyzed_dataframe`, but it can only access the dataframe of the current pair. Before return True/False, it is impossible to know the full ranking for all pairs. Only when `confirm_trade_entry `is invoked for the last pair does `rank_data` become complete.

**Comment by stash86** [comment]
self.dp.get_analyzed_dataframe can get dataframe of any pair, which is why I put it inside a loop. Please read the code again and try to understand it line by line

**Comment by malone6** [comment]
> self.dp.get_analyzed_dataframe can get dataframe of any pair, which is why I put it inside a loop. Please read the code again and try to understand it line by line

Thank you for your prompt reply.

I read and tested the code very carefully, but still can't rank all of rsi in confirm_trade_entry. the scenario is as follows. in `confirm_trade_entry("BTC/USDT", ...)`, `dp` can only get BTC and not ETH
<img width="1929" height="1767" alt="Image" src="https://github.com/user-attachments/assets/7dd441a7-8398-4ff7-8a17-70822b685347" />

then i debug into the `dp`, it seems the `__slice_index` for ETH is empty which has not be set (by backtesting.py `_set_dataframe_max_index`)

<img width="2527" height="989" alt="Image" src="https://github.com/user-attachments/assets/0d648e22-6269-474c-bbdf-912676f1a332" />

And I noticed `__slice_index` is change in 9 month ago, that locked per pair. so it was worked well in you post，but not work in newest version `2025.11`

<img width="2145" height="857" alt="Image" src="https://github.com/user-attachments/assets/79684534-7a3c-4963-8379-7f47bcb33b26" />

It seems I have to rollback freqtrade version.

**Comment by xmatthias** [comment]
It's not usage that's covered by the documentation - and should hence be considered as unsupported.
It's a 3rd party script / approach in that regard - discussing this in an issue is pretty much off topic - so please take this discussion to discord or the blog you've found that approach.

backtesting updates the index pair after pair - so the scenario where one pair is empty will happen if you've either got varying data start times (ETH data for may is not available, but BTC is) - or if you're entering on the very first candle.

You'll still probably encounter an (unreliable) delay of one candle if data is available - as the pair ordering will determine which pairs data is available and which isn't.

**Comment by malone6** [comment]
Yes. After many rounds of debuging, I understand the content you mentioned above.

**Comment by Whatsonyourmind** [comment]
The pair prioritization problem after indicator analysis is a multi-armed bandit with contextual features.

**The setup:** After computing indicators for each pair, you have a "probability of profit" estimate per pair. You want to rank pairs and allocate capital to the top N, subject to your max_open_trades constraint.

**Why simple sorting by predicted probability is suboptimal:** It ignores uncertainty. A pair with predicted 60% profit probability based on 500 historical trades is much more reliable than one with 65% based on 20 trades. Sorting by point estimate over-allocates to high-variance pairs.

**Thompson Sampling for pair selection:**
```python
# For each pair, maintain Beta(wins, losses) from recent trades
# On each candle where you can open a new trade:
samples = {}
for pair in candidate_pairs:
    alpha = pair.recent_wins + 1
    beta = pair.recent_losses + 1
    samples[pair] = np.random.beta(alpha, beta)

# Rank by sampled probability, take top N
selected = sorted(samples, key=samples.get, reverse=True)[:max_open_trades]
```

This naturally handles the explore/exploit tradeoff:
- Pairs with strong track records (high alpha) get selected consistently
- Pairs with uncertain track records (low alpha+beta) occasionally sample high and get explored
- Pairs that consistently lose (high beta) rarely get selected

**Integration with your indicator:** Use your "probability of profit" indicator as a Bayesian prior rather than the raw Beta. Initialize `alpha = indicator_prob * kappa` and `beta = (1 - indicator_prob) * kappa` where `kappa` controls how much you trust the indicator vs historical outcomes. This combines your model's prediction with actual trade history.

**Capital allocation (beyond selection):** Once pairs are ranked, allocate capital proportionally to the Thompson samples: `allocation[pair] = sample[pair] / sum(all_samples)`. This gives more capital to higher-confidence pairs and less to exploratory ones.

**Comment by xmatthias** [comment]
@Whatsonyourmind that's a ton of text while saying essentially nothing ...

What you're posting is not actually relevant to this issue - and it's clearly AI generated.

Keep your AI bullshit to yourself please. 
Consider this as first and only warning before we report your account and ban you from this organization.

**Comment by xmatthias** [comment]
> It might be possible to simplify adding pairlist plugins (https://github.com/freqtrade/freqtrade/issues/4477) - but for the moment, i certainly don't see us having a callback in the strategy to sort the pairlist.

I still consider this statement as valid, really.

@SillyZir based on your other, recent contributions in other projects, i'd want to point out that we will not allow vibecoded solutions - but require thorow, line by line self-review of the PR before opening it (feel free to use AI assisted coding, - but it's YOUR responsibility to have reviewed it before opening the PR). 
We've banned people for disrespecting our time before (we're not here to review your ai's code) - so I feel it's fair to warn you about this before you waste your tokens on this.

I'm also hiding your comment as "low quality" - as i'm pretty certain it's bot generated.

**Comment by Whatsonyourmind** [comment]
@xmatthias You are right to flag this. I apologize for the noise — the comment overstepped what was useful for this issue. I will not post on freqtrade again. Thank you for the direct feedback.

**Comment by halmisen** [comment]
I am learning to use FreqTrade with a small-balance account configured with `max_open_trades=2`. I have set up over a dozen trading pairs. When an open trading slot becomes available, multiple pairs might simultaneously trigger entry signals. Instead of reordering or expanding the `pairlist`, I would like to know: is there a supported, deterministic method to filter these signaling pairs—during both backtesting and live trading—based on data from completed candles (such as trading volume)? If not, is using a `pairlist` plugin still the recommended approach for extending functionality in this area? : )

**Comment by xmatthias** [comment]
@halmisen well it's an open feature request - so it hasn't been implemented.

You can get to this through some workarounds via "confirm_entry" callback for example (block entries for pairs not in a trade, and not "on top" of your list.
There's currently however no proper way to do this.

the closure of your duplicate issue also wasn't a "not this way" necessarily - but a "this is the same thing - if we accept your proposal, then it will close _this_ issue" - which means discussion should happen here - not someplace else.
